# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [54]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [55]:
df = pd.read_csv('/home/abhinav/SPRINGBOARD/Guided_2/data/step3_output.csv')
df = df
df.head(5)
df.keys()

Index([u'Unnamed: 0', u'Name', u'state', u'summit_elev', u'vertical_drop',
       u'trams', u'fastEight', u'fastSixes', u'fastQuads', u'quad', u'triple',
       u'double', u'surface', u'total_chairs', u'Runs', u'TerrainParks',
       u'LongestRun_mi', u'SkiableTerrain_ac', u'Snow Making_ac',
       u'daysOpenLastYear', u'yearsOpen', u'averageSnowfall', u'AdultWeekday',
       u'AdultWeekend', u'projectedDaysOpen', u'NightSkiing_ac', u'clusters'],
      dtype='object')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [56]:
#df.select_dtypes(include=['object'])
df.drop(['Unnamed: 0'],axis=1,inplace=True)
df.head(5)

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple    ...     SkiableTerrain_ac  \
0          0          2     2       0    ...                1610.0   
1          0          0     0       0    ...                 640.0   
2          0          0     0       1    ...                  30.0   
3          1          0     2       2    ...                 777.0   
4          0          1     2       3    ...                 800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 26 columns]

In [57]:
dfo=df.select_dtypes(include=['object']) # select object type columns
dfo.drop('Name',axis=1,inplace=True)
dfo.head(5)


/home/abhinav/miniconda3/lib/python2.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


state
0   Alaska
1   Alaska
2   Alaska
3  Arizona
4  Arizona

In [58]:
pd.get_dummies(dfo)
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(dfo)], axis=1)
df.head(5)

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double      ...        \
0          0          2     2       0       0      ...         
1          0          0     0       0       4      ...         
2          0          0     0       1       0      ...         
3          1          0     2       2       1      ...         
4          0          1     2       3       1      ...         

   state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                   0                   0                0           0   
1                   0                   0                0           0   
2                   0                   0                0           0   
3                   0                   0                0           0   
4                   0                   0                0           0   

   state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
0              0               0                 0                    0   
1              0               0                 0                    0   
2              0               0                 0                    0   
3              0               0                 0                    0   
4              0               0                 0                    0   

   state_Wisconsin  state_Wyoming  
0                0              0  
1                0              0  
2                0              0  
3                0              0  
4                0              0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [59]:
df.keys()

Index([u'Name', u'summit_elev', u'vertical_drop', u'trams', u'fastEight',
       u'fastSixes', u'fastQuads', u'quad', u'triple', u'double', u'surface',
       u'total_chairs', u'Runs', u'TerrainParks', u'LongestRun_mi',
       u'SkiableTerrain_ac', u'Snow Making_ac', u'daysOpenLastYear',
       u'yearsOpen', u'averageSnowfall', u'AdultWeekday', u'AdultWeekend',
       u'projectedDaysOpen', u'NightSkiing_ac', u'clusters', u'state_Alaska',
       u'state_Arizona', u'state_California', u'state_Colorado',
       u'state_Connecticut', u'state_Idaho', u'state_Illinois',
       u'state_Indiana', u'state_Iowa', u'state_Maine', u'state_Maryland',
       u'state_Massachusetts', u'state_Michigan', u'state_Minnesota',
       u'state_Missouri', u'state_Montana', u'state_Nevada',
       u'state_New Hampshire', u'state_New Jersey', u'state_New Mexico',
       u'state_New York', u'state_North Carolina', u'state_Ohio',
       u'state_Oregon', u'state_Pennsylvania', u'state_Rhode Island',
       u'state

In [60]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']
X
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [61]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [62]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [63]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [64]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
from sklearn.metrics import mean_absolute_error, explained_variance_score
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
eval_error = mean_absolute_error(y_test,y_pred)
eval_var = explained_variance_score(y_test,y_pred)
print(eval_error,eval_var)

(5.016922136248756, 0.9384515862520051)


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [65]:
print("This is the intercept",model.intercept_)

('This is the intercept', 64.09946851323181)


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [69]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefficients = pd.DataFrame((abs(lm.coef_)), X.columns, columns=['Coefficient'])
coefficients.sort_values(by='Coefficient',ascending=False)



Coefficient
total_chairs          2.933693e+13
fastQuads             1.112169e+13
surface               1.042019e+13
double                9.182661e+12
triple                8.191565e+12
quad                  6.638960e+12
fastSixes             3.297032e+12
trams                 2.832902e+12
state_New York        2.123032e+12
state_Michigan        2.003566e+12
state_Colorado        1.765258e+12
state_California      1.727469e+12
state_Pennsylvania    1.648460e+12
state_New Hampshire   1.520009e+12
state_Wisconsin       1.520009e+12
state_Vermont         1.474084e+12
state_Minnesota       1.426359e+12
state_Utah            1.376647e+12
state_Idaho           1.324725e+12
state_Montana         1.324725e+12
state_Massachusetts   1.270320e+12
state_Washington      1.213099e+12
state_Oregon          1.213099e+12
state_Maine           1.152644e+12
state_New Mexico      1.152644e+12
state_Wyoming         1.088415e+12
state_North Carolina  9.455174e+11
state_Connecticut     8.644663e+11
state_Ohio            8.644663e+11
state_West Virginia   7.743908e+11
state_Virginia        7.743908e+11
state_Illinois        7.743908e+11
state_Nevada          7.743908e+11
state_Iowa            6.716699e+11
state_Alaska          6.716699e+11
state_New Jersey      5.492541e+11
state_Missouri        5.492541e+11
state_Arizona         5.492541e+11
state_Indiana         5.492541e+11
state_South Dakota    5.492541e+11
state_Tennessee       3.889729e+11
state_Maryland        3.889729e+11
state_Rhode Island    3.889729e+11
fastEight             2.785020e+11
AdultWeekday          1.924292e+01
summit_elev           1.664636e+00
Snow Making_ac        9.436035e-01
Runs                  8.825684e-01
vertical_drop         8.817785e-01
NightSkiing_ac        6.950684e-01
clusters              6.171875e-01
daysOpenLastYear      5.986328e-01
LongestRun_mi         5.668945e-01
yearsOpen             4.604492e-01
SkiableTerrain_ac     4.433594e-01
projectedDaysOpen     4.316406e-01
TerrainParks          3.233948e-01
averageSnowfall       1.162109e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [70]:
df_state = df.drop(list(df.filter(regex = 'state_')), axis = 1)#select columns using regular expression
df_state.head(5)

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double    ...     SkiableTerrain_ac  \
0          0          2     2       0       0    ...                1610.0   
1          0          0     0       0       4    ...                 640.0   
2          0          0     0       1       0    ...                  30.0   
3          1          0     2       2       1    ...                 777.0   
4          0          1     2       3       1    ...                 800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 25 columns]

In [71]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X_ = df_state.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y_ = df_state['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler_ = preprocessing.StandardScaler().fit(X_)
X_
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled_=scaler_.transform(X_) 
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y_ = y_.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train_, X_test_, y_train_, y_test_ = train_test_split(X_scaled_, y_, test_size=0.25, random_state=1)
lm_ = linear_model.LinearRegression()
model_ = lm_.fit(X_train_,y_train_)
y_pred_ = model_.predict(X_test_)
eval_error_ = mean_absolute_error(y_test_,y_pred_)
eval_var_ = explained_variance_score(y_test_,y_pred_)
print(eval_error_,eval_var_)
print("This is the intercept",model_.intercept_)
Coefficient_ = pd.DataFrame((abs(lm_.coef_)), X_.columns, columns=['Coefficient'])
Coefficient_.sort_values(by='Coefficient',ascending=False)

(5.2565241100243085, 0.9311345996558178)
('This is the intercept', 64.07111435118097)


Coefficient
AdultWeekday         19.537445
clusters              5.939183
summit_elev           4.685934
SkiableTerrain_ac     2.202966
surface               1.483651
triple                1.451105
Runs                  1.378389
vertical_drop         1.331524
fastQuads             1.257959
NightSkiing_ac        0.902605
quad                  0.870307
projectedDaysOpen     0.740774
averageSnowfall       0.697219
total_chairs          0.677171
fastSixes             0.594991
LongestRun_mi         0.555266
fastEight             0.319114
yearsOpen             0.317368
Snow Making_ac        0.268045
double                0.169874
TerrainParks          0.147601
daysOpenLastYear      0.093849
trams                 0.021249

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [72]:
df_Mod3 = df_state.drop('summit_elev',axis=1)
df_Mod3.head(5)

Name  vertical_drop  trams  fastEight  fastSixes  fastQuads  \
0       Alyeska Resort           2500      1        0.0          0          2   
1  Eaglecrest Ski Area           1540      0        0.0          0          0   
2     Hilltop Ski Area            294      0        0.0          0          0   
3     Arizona Snowbowl           2300      0        0.0          1          0   
4  Sunrise Park Resort           1800      0        0.0          0          1   

   quad  triple  double  surface    ...     SkiableTerrain_ac  Snow Making_ac  \
0     2       0       0        2    ...                1610.0           113.0   
1     0       0       4        0    ...                 640.0            60.0   
2     0       1       0        2    ...                  30.0            30.0   
3     2       2       1        2    ...                 777.0           104.0   
4     2       3       1        0    ...                 800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         1  
1               90.0             0.0         1  
2              152.0            30.0         1  
3              122.0             0.0         0  
4              104.0            80.0         0  

[5 rows x 24 columns]

In [83]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X_3 = df_Mod3.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y_3 = df_Mod3['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler_3 = preprocessing.StandardScaler().fit(X_3)
X_3
# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled_3=scaler_3.transform(X_3) 
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y_3 = y_3.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train_3, X_test_3, y_train_3, y_test_3 = train_test_split(X_scaled_3, y_3, test_size=0.25, random_state=1)
lm_3 = linear_model.LinearRegression()
model_3 = lm_3.fit(X_train_3,y_train_3)
y_pred_3 = model_3.predict(X_test_3)
eval_error_3 = mean_absolute_error(y_test_3,y_pred_3)
eval_var_3 = explained_variance_score(y_test_3,y_pred_3)
print(eval_error_3,eval_var_3)
print("This is the intercept",model_3.intercept_)
Coefficient_3 = pd.DataFrame((abs(lm_3.coef_)), X_3.columns, columns=['Coefficient'])
Coefficient_3.sort_values(by='Coefficient',ascending=False)

(5.221013556255185, 0.9319511731655595)
('This is the intercept', 64.0425214534508)


Coefficient
AdultWeekday         19.917284
clusters              2.655862
vertical_drop         2.280169
triple                1.411468
surface               1.207337
Runs                  1.141280
SkiableTerrain_ac     1.080944
quad                  1.005293
fastQuads             0.884289
averageSnowfall       0.862302
projectedDaysOpen     0.822991
total_chairs          0.624404
Snow Making_ac        0.563383
fastSixes             0.557475
NightSkiing_ac        0.495109
LongestRun_mi         0.483432
TerrainParks          0.411585
yearsOpen             0.293034
trams                 0.239692
daysOpenLastYear      0.236875
double                0.178259
fastEight             0.114476

In [0]:
5.221013556255185, 0.9319511731655595

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9384515862520051|  5.016922136248756| 'base_elev'|
| Model 2. | 0.9311345996558178| 5.2565241100243085 |'state', 'base_elev'|
| Model 3. | 0.9319511731655595 |5.221013556255185|'state','summit_elev','base_elev'|

Model Selection:

Model number 3 is the best model. This is because it doesn't rely on the correlated features and features which are not very useful from the management perspective. Further, the model has an overall low value of error and variance.